lib/galaxy/model/dataset_collections/types/list.py

```diff
@@ -1,14 +1,33 @@
+from typing import (
+    Iterable,
+    Mapping,
+    TYPE_CHECKING,
+    Union,
+)
+
 from galaxy.model import DatasetCollectionElement
 from . import BaseDatasetCollectionType
 
+if TYPE_CHECKING:
+    from galaxy.model import (
+        DatasetCollection,
+        HistoryDatasetAssociation,
+        LibraryDatasetDatasetAssociation,
+    )
+
 
 class ListDatasetCollectionType(BaseDatasetCollectionType):
     """A flat list of named elements."""
 
     collection_type = "list"
 
-    def generate_elements(self, elements):
-        for identifier, element in elements.items():
+    def generate_elements(
+        self,
+        dataset_instances: Mapping[
+            str, Union["DatasetCollection", "HistoryDatasetAssociation", "LibraryDatasetDatasetAssociation"]
```

Comment:
Please define a named type for these and reuse in implementations. There are already two and open PRs will add two more and in builder.

```python
DatasetInstanceType = Union["DatasetCollection", "HistoryDatasetAssociation", "LibraryDatasetDatasetAssociation"]
DatasetInstanceMapping = Mapping[str, DatasetInstanceType]
```

My sample sheet PR has some additional, conflicting typing also - https://github.com/galaxyproject/galaxy/pull/19305/commits/dc8bad94a8ffc025ac7968bd9d541a672db89149. Any chance you can merge it in or otherwise I would be more than happy to merge them and make this change to this PR.

Reply:
@jmchilton Thanks for the review!
I've added `DatasetInstanceMapping` as you suggested, and merged https://github.com/galaxyproject/galaxy/commit/dc8bad94a8ffc025ac7968bd9d541a672db89149 into my commit, with some small adjustments, hope that's OK!